flutter/lib/resources/cache_manager.dart

```diff
@@ -0,0 +1,116 @@
+  late final ArchiveCacheHelper archiveCacheHelper;
+  Map<String, String> _resourcesMap = {};
+
+  CacheManager(this.loadedResourcesDir) {
+    fileCacheHelper = FileCacheHelper(loadedResourcesDir);
+    archiveCacheHelper = ArchiveCacheHelper(fileCacheHelper);
+  }
+
+  String? get(String url) {
+    return _resourcesMap[url];
+  }
+
+  Future<void> deleteLoadedResources(List<String> nonRemovableResources,
+      [int atLeastDaysOld = 0]) async {
+    final directory = Directory(loadedResourcesDir);
+
+    // can't use 'await for' here because we delete files,
+    // and 'await for' on stream from directory.list() throws exceptions when file is missing
+    for (final file in await directory.list(recursive: true).toList()) {
+      // skip files in already deleted folders
+      if (!await file.exists()) continue;
+      final relativePath = file.path
+          .substring(loadedResourcesDir.length + 1)
+          .replaceAll('\\', '/');
+      var nonRemovable = false;
+      for (var resource in nonRemovableResources) {
+        // relativePath.startsWith(resource): if we want to preserve a folder resource
+        // resource.startsWith(relativePath): if we want to preserve a file resource
+        //   for example:
+        //   we are checking folder 'github.com'
+        //   resource is 'github.com/mlcommons/mobile_models/raw/main/v0_7/datasets/ade20k'
+        if (relativePath.startsWith(resource) ||
+            resource.startsWith(relativePath)) {
+          nonRemovable = true;
+          break;
+        }
+      }
+      if (nonRemovable) continue;
+      if (atLeastDaysOld > 0) {
+        var stat = await file.stat();
+        if (DateTime.now().difference(stat.modified).inDays < atLeastDaysOld) {
+          continue;
+        }
+      }
+      await file.delete(recursive: true);
+    }
+  }
+
+  Future<void> purgeOutdatedCache(int atLeastDaysOld) async {
+    var currentResources = <String>[];
+    for (var r in _resourcesMap.values) {
+      currentResources.add(r.substring(loadedResourcesDir.length + 1));
+    }
+    return deleteLoadedResources(currentResources, atLeastDaysOld);
+  }
+
+  Future<void> cache(List<String> urls, void Function(double) reportProgress,
+      bool purgeOldCache) async {
+    final resourcesToDownload = <String>[];
+    _resourcesMap = {};
+
+    for (final resource in urls) {
+      if (_resourcesMap.containsKey(resource)) continue;
+      if (resourcesToDownload.contains(resource)) continue;
+
+      String path;
+      if (isResourceAnArchive(resource)) {
+        path = await archiveCacheHelper.get(resource, false);
+      } else {
+        path = await fileCacheHelper.get(resource, false);
+      }
+
+      if (path != '') {
+        _resourcesMap[resource] = path;
+        continue;
+      }
+
+      resourcesToDownload.add(resource);
+
+      continue;
+    }
+    await _download(resourcesToDownload, reportProgress);
+
+    if (purgeOldCache) {
+      const atLeastDaysOld = 30 * 9;
```

Comment:
Maybe we can make this `atLeastDaysOld` a global constant or at least class constant.

Reply:
fixed